python/cuml/manifold/simpl_set.pyx

```diff
@@ -0,0 +1,367 @@
+    X: array of shape (n_samples, n_features)
+        The data to be modelled as a fuzzy simplicial set.
+    n_neighbors: int
+        The number of neighbors to use to approximate geodesic distance.
+        Larger numbers induce more global estimates of the manifold that can
+        miss finer detail, while smaller values will focus on fine manifold
+        structure to the detriment of the larger picture.
+    random_state: numpy RandomState or equivalent
+        A state capable being used as a numpy random state.
+    metric: string or function (optional, default 'euclidean')
+        unused
+    metric_kwds: dict (optional, default {})
+        unused
+    set_op_mix_ratio: float (optional, default 1.0)
+        Interpolate between (fuzzy) union and intersection as the set operation
+        used to combine local fuzzy simplicial sets to obtain a global fuzzy
+        simplicial sets. Both fuzzy set operations use the product t-norm.
+        The value of this parameter should be between 0.0 and 1.0; a value of
+        1.0 will use a pure fuzzy union, while 0.0 will use a pure fuzzy
+        intersection.
+    local_connectivity: int (optional, default 1)
+        The local connectivity required -- i.e. the number of nearest
+        neighbors that should be assumed to be connected at a local level.
+        The higher this value the more connected the manifold becomes
+        locally. In practice this should be not more than the local intrinsic
+        dimension of the manifold.
+    verbose: bool (optional, default False)
+        Whether to report information on the current progress of the algorithm.
+    Returns
+    -------
+    fuzzy_simplicial_set: coo_matrix
+        A fuzzy simplicial set represented as a sparse matrix. The (i,
+        j) entry of the matrix represents the membership strength of the
+        1-simplex between the ith and jth sample points.
+    """
+
+    deterministic = random_state is not None
+    if not isinstance(random_state, int):
+        if isinstance(random_state, np.random.RandomState):
+            rs = random_state
+        else:
+            rs = np.random.RandomState(random_state)
+        random_state = rs.randint(low=np.iinfo(np.int32).min,
+                                  high=np.iinfo(np.int32).max,
+                                  dtype=np.int32)
+
+    cdef UMAPParams* umap_params = new UMAPParams()
+    umap_params.n_neighbors = <int> n_neighbors
+    umap_params.random_state = <uint64_t> random_state
+    umap_params.deterministic = <bool> deterministic
+    umap_params.set_op_mix_ratio = <float> set_op_mix_ratio
+    umap_params.local_connectivity = <float> local_connectivity
+    umap_params.verbosity = <int> verbose
+
+    X_m, n_rows, n_cols, dtype = \
+        input_to_cuml_array(X, order='C', check_dtype=np.float32)
+
+    handle = Handle()
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+    cdef unique_ptr[COO] fss_graph = get_graph(handle_[0],
```

Comment:
See comment below about potentially attaching the COO object to the cupy sparse object so that it can be cleaned up at the proper time (and eliminate the copy).